File: api_nestjs_distributed_order_platform/src/application/use-case/stock/update-stock.use-case.ts
import { ConflictException, Inject, Injectable, NotFoundException } from "@nestjs/common";
import type { StockRepositoryInterface } from "src/application/port/stock.repository";
import { StockEntity } from "src/domain/entities/stock.entity";
import { UpdateStockInput } from "src/shared/interfaces/stock.interface";
import { STOCK_REPOSITORY } from "src/shared/tokens_nest/stock.token";

@Injectable()
export class UpdateStockUseCase {
    constructor(
        @Inject(STOCK_REPOSITORY)
        private readonly stockRepository: StockRepositoryInterface,
    ) {}

    async execute(input: UpdateStockInput): Promise<StockEntity> {
        const stock = await this.stockRepository.findByProductId(input.productId);
        if (!stock) {
            throw new NotFoundException('Stock not found');
        }

        const expectedVersion = stock.getVersion();
        stock.setQuantities(
            input.availableQuantity,
            input.reservedQuantity ?? stock.getReservedQuantity(),
        );

        const updated = await this.stockRepository.update(stock, expectedVersion);
        if (!updated) {
            throw new ConflictException('Stock was modified by another request');
        }

        return stock;
    }
}

File: api_java_distributed_order_platform/src/main/java/com/project/order/application/usecase/stock/UpdateStockUseCase.java
package com.project.order.application.usecase.stock;

import com.project.order.application.exception.ConflictException;
import com.project.order.application.exception.NotFoundException;
import com.project.order.application.port.StockRepository;
import com.project.order.domain.model.Stock;
import org.springframework.dao.OptimisticLockingFailureException;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

/**
 * Ajuste administrativo direto de quantidades (PUT /stock/{productId}).
 * A concorrência é resolvida pelo @Version do Stock na camada de
 * infraestrutura: se outra requisição alterou o registro entre o findById
 * e o update, o adapter JPA lança OptimisticLockingFailureException e aqui
 * a gente traduz isso pra um 409 de domínio.
 */
@Service
public class UpdateStockUseCase {

    private final StockRepository stockRepository;

    public UpdateStockUseCase(StockRepository stockRepository) {
        this.stockRepository = stockRepository;
    }

    @Transactional
    public Stock execute(String productId, int availableQuantity, Integer reservedQuantity) {
        Stock stock = stockRepository.findById(productId)
                .orElseThrow(() -> new NotFoundException("Stock not found"));

        int nextReserved = reservedQuantity != null ? reservedQuantity : stock.getReservedQuantity();
        stock.setQuantities(availableQuantity, nextReserved);

        try {
            return stockRepository.update(stock);
        } catch (OptimisticLockingFailureException e) {
            throw new ConflictException("Stock was modified by another request");
        }
    }
}
